File: frontend/src/services/scanService.js
import api from './api'

/**
 * Scan Service
 * ------------
 * Handles QR scan operations and scan logs.
 */

const scanService = {
  /**
   * Process a QR code scan
   * @param {number} areaId - Area ID
   * @param {string} type - 'ENTRY' or 'EXIT'
   * @returns {Promise} - Scan result with updated count
   */
  processScan: async (areaId, type) => {
    const response = await api.post('/scans', { areaId, scanType: type })
    return response.data
  },

  /**
   * Get recent scan logs
   * @param {number} limit - Number of logs to fetch (default 50)
   * @returns {Promise} - List of recent scans
   */
  getRecentScans: async (limit = 50) => {
    const response = await api.get(`/scans/recent?limit=${limit}`)
    return response.data
  },

  /**
   * Get scan logs for a specific area
   * @param {number} areaId - Area ID
   * @returns {Promise} - List of scans for the area
   */
  getScansByArea: async (areaId) => {
    const response = await api.get(`/scans/area/${areaId}`)
    return response.data
  },

  /**
   * Get scan logs for today
   * @returns {Promise} - Today's scan logs
   */
  getTodayScans: async () => {
    const response = await api.get('/scans/today')
    return response.data
  },

  /**
   * Register a scan (alias for processScan)
   * @param {number} areaId - Area ID
   * @param {string} type - 'ENTRY' or 'EXIT'
   */
  registerScan: async function(areaId, type) {
    return this.processScan(areaId, type)
  },

  /**
   * Get hourly trend data for an area
   * @param {number} areaId - Area ID
   * @returns {Promise} - Hourly trend data
   */
  getHourlyTrend: async (areaId) => {
    try {
      const response = await api.get(`/scans/area/${areaId}/trend`)
      return response.data
    } catch (err) {
      // Return mock data if API not available
      return generateMockTrend()
    }
  },
}

// Helper function for mock trend data
function generateMockTrend() {
  const hours = []
  const now = new Date()
  for (let i = 12; i >= 0; i--) {
    const hour = new Date(now - i * 3600000)
    hours.push({
      hour: hour.getHours() + ':00',
      entries: Math.floor(Math.random() * 50) + 10,
      exits: Math.floor(Math.random() * 40) + 5,
      count: Math.floor(Math.random() * 100) + 20
    })
  }
  return hours
}

export default scanService

File: backend/src/main/java/com/crowdmanagement/service/ScanService.java
package com.crowdmanagement.service;

import com.crowdmanagement.dto.ScanRequest;
import com.crowdmanagement.dto.ScanResponse;
import com.crowdmanagement.dto.AreaResponse;
import com.crowdmanagement.entity.Area;
import com.crowdmanagement.entity.ScanLog;
import com.crowdmanagement.entity.ScanType;
import com.crowdmanagement.repository.ScanLogRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Scan Service
 * ------------
 * Handles QR code scan operations and scan log management.
 */
@Service
public class ScanService {

    @Autowired
    private ScanLogRepository scanLogRepository;

    @Autowired
    private AreaService areaService;

    @Autowired
    private WebSocketService webSocketService;

    @Autowired
    private AlertService alertService;

    /**
     * Process a QR code scan (entry or exit)
     * @param request Scan details
     * @return Scan result with updated count
     */
    @Transactional
    public ScanResponse processScan(ScanRequest request) {
        // Get the area
        Area area = areaService.getAreaEntityById(request.getAreaId());

        // Update the count based on scan type
        Integer newCount;
        if (request.getScanType() == ScanType.ENTRY) {
            newCount = areaService.incrementCount(request.getAreaId());
        } else {
            newCount = areaService.decrementCount(request.getAreaId());
        }

        // Create scan log
        ScanLog scanLog = new ScanLog(area, request.getScanType());
        scanLog = scanLogRepository.save(scanLog);

        // Refresh area entity to get updated count
        area = areaService.getAreaEntityById(request.getAreaId());

        // Check and generate alerts if thresholds are breached
        alertService.checkAndGenerateAlert(area, request.getScanType());

        // Broadcast real-time update via WebSocket
        AreaResponse updatedArea = areaService.getAreaByIdPublic(request.getAreaId());
        webSocketService.broadcastAreaUpdate(updatedArea);
        webSocketService.broadcastScanEvent(request.getAreaId(), request.getScanType().name(), newCount);

        return ScanResponse.fromEntity(scanLog, newCount);
    }

    /**
     * Get recent scan logs
     * @param limit Number of logs to return
     * @return List of recent scans
     */
    public List<ScanResponse> getRecentScans(int limit) {
        return scanLogRepository.findRecentScans(limit)
                .stream()
                .map(log -> ScanResponse.fromEntity(log, log.getArea().getCurrentCount()))
                .collect(Collectors.toList());
    }

    /**
     * Get scans for a specific area
     * @param areaId Area ID
     * @return List of scans for the area
     */
    public List<ScanResponse> getScansByArea(Long areaId) {
        return scanLogRepository.findByAreaIdOrderByTimestampDesc(areaId)
                .stream()
                .map(log -> ScanResponse.fromEntity(log, log.getArea().getCurrentCount()))
                .collect(Collectors.toList());
    }

    /**
     * Get today's scans
     * @return List of today's scans
     */
    public List<ScanResponse> getTodayScans() {
        LocalDateTime startOfDay = LocalDate.now().atStartOfDay();
        LocalDateTime endOfDay = LocalDate.now().atTime(LocalTime.MAX);

        return scanLogRepository.findByTimestampBetweenOrderByTimestampDesc(startOfDay, endOfDay)
                .stream()
                .map(log -> ScanResponse.fromEntity(log, log.getArea().getCurrentCount()))
                .collect(Collectors.toList());
    }

    /**
     * Get hourly trend data for an area
     * @param areaId Area ID
     * @return Hourly scan counts
     */
    public List<HourlyTrendData> getHourlyTrend(Long areaId) {
        LocalDateTime startOfDay = LocalDate.now().atStartOfDay();
        LocalDateTime endOfDay = LocalDate.now().atTime(LocalTime.MAX);

        List<Object[]> rawData = scanLogRepository.getHourlyScanCounts(areaId, startOfDay, endOfDay);
        
        List<HourlyTrendData> result = new ArrayList<>();
        for (Object[] row : rawData) {
            Integer hour = (Integer) row[0];
            Long entries = (Long) row[1];
            Long exits = (Long) row[2];
            result.add(new HourlyTrendData(
                    hour + ":00",
                    entries.intValue(),
                    exits.intValue(),
                    entries.intValue() - exits.intValue()
            ));
        }
        return result;
    }

    /**
     * Inner class for hourly trend data
     */
    public static class HourlyTrendData {
        public String hour;
        public Integer entries;
        public Integer exits;
        public Integer count;

        public HourlyTrendData(String hour, Integer entries, Integer exits, Integer count) {
            this.hour = hour;
            this.entries = entries;
            this.exits = exits;
            this.count = count;
        }
    }
}
